utils_test.py

```diff
@@ -251,3 +251,141 @@ def test_get_hashable_value(self):
                 utils.get_hashable_value(json1_deepcopy),
                 utils.get_hashable_value(json2_deepcopy),
             })
+
+    def test_convert_to_text_angular(self):
```

Comment:
OK -- so, these tests need to be really comprehensive, since this is high-risk code. Here is what I suggest:

- Group the observed and expected output together, so that they are easy for a developer reading the code to compare.
- For every single problematic issue you're spotting at the moment on the server, add a test case for it here. Then run a for loop over these cases and check the following, so that we have assurance that after the run, everything is in a good state:
  - Before it's converted, the HTML fails validation.
  - The converted HTML matches the expected HTML.
  - The expected HTML passes validation.

I expect that there will end up being quite a lot of examples here. We'll need to manually review all of them before this PR goes in.

Also, add a test somewhere (maybe exp_domain.py) where you create a dummy exploration with an older schema version whose content fields are several of the problematic HTMLs. Retrieve it (upgrading it to the latest schema version in the process) and check that it passes the textAngular validation.

Reply:
Okay will do!

Reply:
Oh, one other thought. It might be worth generating the yaml for a dummy exploration whose content fields are populated with the problematic HTML. During testing we can then run through and manually inspect it, and ensure that everything at least _looks_ OK.

I feel like, if we do these checks, we can be very certain that there are no issues from either the infrastructure or UX sides -- and safety is paramount in this sort of job. If you think of any other checks, let us know!

Reply:
Hi @seanlip, for testing the three points as mentioned by you: 
- Before it's converted, the HTML fails validation.
- The converted HTML matches the expected HTML.
- The expected HTML passes validation
Since the test will use two jobs - migration and validation, should some separate test be written for this?

Reply:
Yes, it's fine to write these tests in a separate suite if you like. I don't mind too much either way because all these are one-off jobs, right? So exp_jobs_one_off_test.py seems like a reasonable location...

Reply:
OK, its fine to write them in exp_jobs_one_off.py, will do! Thanks :smile:

Reply:
I have added the tests, PTAL!

Reply:
It seems strange to put the test data in feconf. It's not a config property and not relevant to the non-test code.

We have a location core/tests/data for test data. You could put it there, perhaps in a json file which you then extract here.

Also, I don't see the test checking that verifies that the "before edit" case for each of these fails validation, and the "after edit" case passes validation.

Reply:
Changed the location of test data. Also, the before edit and after edit is added in `exp_jobs_one_off_test.py`, I have added the comments for the same.